include/quicr/detail/messages.h

```diff
@@ -14,119 +14,451 @@
+                case StreamHeaderType::kSubgroupExplicitNotEndOfGroupNoExtensions:
+                case StreamHeaderType::kSubgroupExplicitNotEndOfGroupWithExtensions:
+                case StreamHeaderType::kSubgroupExplicitEndOfGroupNoExtensions:
+                case StreamHeaderType::kSubgroupExplicitEndOfGroupWithExtensions:
+                    return SubgroupIdType::kExplicit;
+            }
+            assert(false);
+            return SubgroupIdType::kIsZero;
+        }
+    };
+
+    /**
+     * Describes the properties of a datagram header type.
+     */
+    struct DatagramHeaderProperties
+    {
+        // True if this object is end of the group.
+        const bool end_of_group;
+        // True if this object has extensions.
+        const bool has_extensions;
+
+        constexpr explicit DatagramHeaderProperties(const DatagramHeaderType type)
+          : end_of_group(EndOfGroup(type))
+          , has_extensions(HasExtensions(type))
+        {
+        }
+
+        constexpr DatagramHeaderProperties(const bool end_of_group, const bool has_extensions)
+          : end_of_group(end_of_group)
+          , has_extensions(has_extensions)
+        {
+        }
+
+        /**
+         * Get the type of this datagram header based on its properties.
+         * @return The DatagramHeaderType corresponding to these properties.
+         */
+        constexpr DatagramHeaderType GetType() const
+        {
+            if (end_of_group) {
+                return has_extensions ? DatagramHeaderType::kEndOfGroupWithExtensions
+                                      : DatagramHeaderType::kEndOfGroupNoExtensions;
+            }
+            return has_extensions ? DatagramHeaderType::kNotEndOfGroupWithExtensions
+                                  : DatagramHeaderType::kNotEndOfGroupNoExtensions;
+        }
+
+      private:
+        static constexpr bool EndOfGroup(const DatagramHeaderType type)
+        {
+            switch (type) {
+                case DatagramHeaderType::kEndOfGroupNoExtensions:
+                case DatagramHeaderType::kEndOfGroupWithExtensions:
+                    return true;
+                case DatagramHeaderType::kNotEndOfGroupNoExtensions:
+                case DatagramHeaderType::kNotEndOfGroupWithExtensions:
+                    return false;
+            }
+            assert(false);
+            return false;
+        }
+
+        static constexpr bool HasExtensions(const DatagramHeaderType type)
+        {
+            switch (type) {
+                case DatagramHeaderType::kNotEndOfGroupNoExtensions:
+                case DatagramHeaderType::kEndOfGroupNoExtensions:
+                    return false;
+                case DatagramHeaderType::kNotEndOfGroupWithExtensions:
+                case DatagramHeaderType::kEndOfGroupWithExtensions:
+                    return true;
+            }
+            assert(false);
+            return false;
```

Comment:
This might be condensed to
```suggestion
return type == DatagramHeaderType::kNotEndOfGroupWithExtensions || type == DatagramHeaderType::kEndOfGroupWithExtensions;
```

Reply:
I wrote some stuff about this in the PR comments - I am intentionally not doing this.